### PR TITLE
Fixes #4 - Define a constant instead of duplicating this literal "robots" 3 times.

### LINKED ISSUE #4: Define a constant instead of duplicating this literal "robots" 3 times.
### Issue Type: Technical Debt

[SonarCloud Issue ](https://sonarcloud.io/project/issues?assignees=Shine-23-AcPPV%40github&directories=robocode.battle%2Fsrc%2Fmain%2Fjava%2Fnet%2Fsf%2Frobocode%2Fbattle%2Crobocode.battle%2Fsrc%2Fmain%2Fjava%2Fnet%2Fsf%2Frobocode%2Fbattle%2Fpeer%2Crobocode.battle%2Fsrc%2Fmain%2Fjava%2Fnet%2Fsf%2Frobocode%2Fbattle%2Fsnapshot%2Crobocode.battle%2Fsrc%2Fmain%2Fjava%2Fnet%2Fsf%2Frobocode%2Frecording&issueStatuses=OPEN%2CCONFIRMED&id=soen6431group10_robocodeFall2024&open=AZIuy4b_SndwZUJuspPm)
File: `robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java`
Literal: `robots`
Number of times duplicated: 3 
SonarQube Rule: [java - String literals should not be duplicated](https://sonarcloud.io/organizations/soen6431group10/rules?open=java%3AS3776&rule_key=java%3AS3776)

### Description

The literal `robots` appears three times in the code. To improve maintainability and reduce redundancy, it must be defined as a constant. Duplicating string literals can lead to inconsistencies and complicate future updates.

### Impact on Software Qualities

Maintainability: Duplicating string literals reduces code readability by risking inconsistencies and complicating updates.

### Proposed Solution
Define a constant to replace the repeated literal "robots" that appears three times. Consider the following steps:

- Identify Duplicates: Review the code to locate all instances where the literal string `robots` appears.
- Declare a Constant: Define a constant at the class level to store the string `robots`.
- Replace Duplicates: Update the method to replace each instance of the string `robots` with the newly defined constant `ROBOTS`.

### Acceptance Criteria

- A constant named `ROBOTS` is defined at the class level to store the string `robots`.
- All instances of the literal `robots` in the file are replaced with the `ROBOTS` constant.

### Steps to Reproduce

1. Run a SonarQube analysis on the project.
2. Observe the Code Smell report for `robots` duplicated literal.

### Additional Information

Resolving this issue will enhance the maintainability of the codebase and simplify future development efforts with the `TurnSnapshot` class for developers.

### PR DESCRIPTION
**Pull request by Shine-23**
## Fixes #4 - Define a constant instead of duplicating this literal "robots" in TurnSnapshot.java

[SonarCloud Issue Link](https://sonarcloud.io/project/issues?assignees=Shine-23-AcPPV%40github&issueStatuses=OPEN%2CCONFIRMED&id=soen6431group10_robocodeFall2024&open=AZIuy4b_SndwZUJuspPm)

The pull request removes the literal `robots` in TurnSnapshot.java by defining a constant `ROBOT`. This change aims to avoid duplication and improve project's readability and maintainability.

## Key Changes:

- `ROBOT`: Define a constant to perform all the same string replacement.
- `robots`: Remove all the redundant and identical string.

## Benefits:

- Reduced Cognitive Load: Duplicated string literals make the process of refactoring complex and error-prone, as any change would need to be propagated on all occurrences.
- Improved Maintainability: Isolating specific functionality into well-named constants allows for easier updates and modifications in the future.

## Conclusion:
This refactoring successfully address duplicated problem of `robots`, which improves maintainability and readability.

### CODE REVIEW

**Comment by 0matleb2** [comment]
Double check your commits on the PR. It's showing there are no changes.

**Comment by 0matleb2** [comment]
Looks like you've referenced the wrong issue in your commits. The issue you should reference is #4.

**Review by 0matleb2 — CHANGES_REQUESTED** [review]
There are no changes in the PR. Double check your commits @Shine-23.

Also it looks like you references the wrong issue in your commit message and PR title. Should be issue #4.

File: robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java [hunk 1]
```diff
@@ -181,32 +184,36 @@ public void writeXml(XmlWriter writer, SerializableOptions options) throws IOExc
 				writer.writeAttribute("ver", serialVersionUID);
 			}
 
-			writer.startElement(options.shortAttributes ? "rs" : "robots"); {
```
Comment (0matleb2) [inline robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:184]: The open brace after `writer.startElement()` defines a new scope, but it looks like you've removed it which has an impact on the scope of the variables that were previously in this block. This is a change that isn't related to issue #4 so I'd suggest reverting it.

File: robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java [hunk 2]
```diff
@@ -34,6 +34,9 @@ public final class TurnSnapshot implements java.io.Serializable, IXmlSerializabl
 
 	private static final long serialVersionUID = 1L;
 
+	/** Define a constant for the "robots" literal */
```
Comment (0matleb2) [inline robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:37]: This comment is redundant and can be removed.

File: robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java [hunk 3]
```diff
@@ -34,6 +34,9 @@ public final class TurnSnapshot implements java.io.Serializable, IXmlSerializabl
...
+	private static final String ROBOT = "robots";
```
Comment (0matleb2) [inline robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:38]: I'd suggest a more descriptive variable name.
```suggestion
	private static final String ROBOTS_XML_ELEMENT_NAME = "robots";
```

**Review by 0matleb2 — CHANGES_REQUESTED** [review]
Extracting the duplicated string used for the XML element name into a constant is a superficial resolution IMO, but one that I'd accept. For me, it points to a need for a much larger scale refactoring to extract classes for reading and writing turn snapshot XML files.

File: robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java (hunk 1 above)
Comment (Shine-23) [reply robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:184]: If I revert this I'm getting another issue related to this.

Comment (0matleb2) [reply robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:184]: What issue?

Comment (Shine-23) [reply robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:184]: Something like this "Nested code blocks should not be used"

Comment (0matleb2) [reply robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:184]: [This one](https://sonarcloud.io/project/issues?open=AZIuy4b_SndwZUJuspPj&id=soen6431group10_robocodeFall2024). It's a separate issue so we shouldn't try to fix it in the same PR as #4.

Comment (Shine-23) [reply robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:184]: Should I revert my changes

Comment (0matleb2) [reply robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:184]: Yes, revert the changes to the nested code block.

Comment (Shine-23) [reply robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:184]: Once, I revert it back, my sonar analysis is failing and show this issue

Comment (0matleb2) [reply robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:184]: I marked it as an [accepted issue](https://sonarcloud.io/project/issues?open=AZIuy4b_SndwZUJuspPj&id=soen6431group10_robocodeFall2024) in SonarCloud, to say that it is a valid issue, but will be fixed later. The quality gate code check is now passing!

Comment (Shine-23) [reply robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:184]: ok !

**Review by 0matleb2 — APPROVED** [review]
LGTM! Thanks for the changes @Shine-23.

**Review by LeeELIOT — APPROVED** [review]

**Review by sakshirk — APPROVED** [review]
LGTM!

File: robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java [hunk 4]
```diff
@@ -34,6 +34,8 @@ public final class TurnSnapshot implements java.io.Serializable, IXmlSerializabl
 
 	private static final long serialVersionUID = 1L;
 
+	private static final String  ROBOTS_XML_ELEMENT_NAME_STRING = "robots";
```
Comment (0matleb2) [inline robocode.battle/src/main/java/net/sf/robocode/battle/snapshot/TurnSnapshot.java:37]: There's a whitespace issue on this line. I also suggest removing "string" from the variable name, it's redundant.

**Review by 0matleb2 — CHANGES_REQUESTED** [review]

**Pull request merged**